### PR TITLE
bugfix(graphql): Fix single line and empty description formatting

### LINKED ISSUE #12607: Prettier GraphQL comment leading space bug
**Prettier 2.6.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAdNGAEApCALKTAEQjk3QrVgE8AHMgBTgCcBnaTYVQzKAQwC2cJJgDKMZgEsoAc24BfEABoQEWjEnRWyUH2bMIAdwZ6E2lHwA2hvtW0qARsz5gA1nBijBcADLS4yABmVqxwjs5uHqK0LtIyyBIArmEgcAIOcAAmmVk+fLKJfDJwAGIQzAJ8MBqyyCB8iTAQyiB4MAKWAOp4kvCsMWBwoma9kgBuvdR1YKz2INKhzDAMzjKVQSEpAFasAB6icZZwAIqJEPAblqEqMWwsdTLOtHgAjpYttFKwnZKZMHjIAAcAAYbgZQp0nnVPnBFmMAioXmd4Cs1OZ6qwALRQOBZLItZhwJGSQkrIrrJDBK4pUICSQJZjJFSsQ4nZEBSmbFQwPgOH5-AFIABM3OckkscQAwhABBTUqwAKwtRKhAAqvPMVOuIDGyQAklAcrBRGApOoAIKG8TUI6XULyeRAA)

```sh
# Options (if any):
--parser graphql
```

**Input:**

```gql
""" John Doe """
type Person {
  name: String
}
```

**Output:**

```gql
"""
 John Doe
"""
type Person {
  name: String
}
```

**Expected behavior:**

```gql
"""
John Doe
"""
type Person {
  name: String
}
```

Seems like the leading space in the comment is not removed.
Running `prettier` twice (2x) will yield the expected result.

Also created a tiny repo with this bug: https://github.com/chimurai/prettier-graphql-comment-bug

### PR DESCRIPTION
**Pull request by chimurai**
## Description

~GraphQL `description` markdown formatting (https://spec.graphql.org/October2021/#sec-Descriptions)~

fixes #12607

## Checklist
 

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/graphql/12608.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by chimurai** [comment]
Not sure why only [Node.js 16 on ubuntu-latest](https://github.com/prettier/prettier/runs/5860647570?check_suite_focus=true#logs) failed.

File: src/language-graphql/printer-graphql.js [hunk 1]
```diff
@@ -128,7 +128,7 @@ function genericPrint(path, options, print) {
         return [
           '"""',
           hardline,
-          join(hardline, node.value.replace(/"""/g, "\\$&").split("\n")),
```
Comment (fisker) [inline src/language-graphql/printer-graphql.js:131]: I think it's a bug in parser, check the `StringValue.value` the leading space are removed, or is there a spec says it should be removed?

**Prettier 2.6.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAdNHVQAS4MICuAzjBALZwBOWu2hJ5VW6LasAngA5zYAKVRaNmA1sUAIYUk2AMoxKASygBzLAF9mUVixy16pCtV11iBplsxsYXHv0qCcI3RKmz5S1VDUgANCAicMArQRMig4pSUEADuvBEIoSjiADbR4uyhfgBGlOJgANZwMDKScAAySnDIAGYpRHDZuQVFMpx5HsjyBA0gcGRZcAAmg0Nl4ioE4spwAGIQlGTiMEEqyCDiBKS+IAAWMGTJAOo7CvBEbWBwMgmnCgBup+xrYESZIEr1lDC8ucqLNXUegArIgADxkHmScAAigQIPAAcl6n42vYqGtlLlODsAI7JbacRSwQ4KQYwHbIAAcAAYUVF6ocsWtCXBPncqn4cXD4D8Aol1kQALRQOBDIbbShwLkKSU-Kb-JC1JE9epkBSdSjdPxESEw7lVRWAvwwcRZElkilIABMxtyCmSHjw5AVvSIAFZtsQ4AAVU2JJXIkB3boASSgI1gMjAikCAEFw3J2FDEfU1GogA)

```sh
--parser graphql
```

**Input:**

```graphql
"""
   Customer
   Customer
"""
type Person {
  name: String
}

"""
    Customer
   Customer
"""
type Person {
  name: String
}
```

**Output:**

```graphql
"""
Customer
Customer
"""
type Person {
  name: String
}

"""
 Customer
Customer
"""
type Person {
  name: String
}

```

**Review by fisker — CHANGES_REQUESTED** [review]

File: src/language-graphql/printer-graphql.js (hunk 1 above)
Comment (fisker) [reply src/language-graphql/printer-graphql.js:131]: I'm not familiar with graphql, why it's called `StringValue` not `Comment` or something?

Comment (fisker) [reply src/language-graphql/printer-graphql.js:131]: Ah, maybe I start understanding, it's a `description`, so may be we should only apply logic to the `StringValue ` in `description`?

**Comment by fisker** [comment]
I think this solution breaks case like this

**Prettier pr-12608**
[Playground link](https://deploy-preview-12608--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAdNHVQATYLTYA2AlgM4xa4EnlZwAe8AhtpALZsIVTq9qwBPAA5xsABTgAnUtGzBK2KE05JsAZRiTiUAOZYAviAA0ICEJjFopZKCaTJEAO5i7CayiaFHTAdZMAjSSYwAGs4GDVlOAAZbThkADNPUjgAoNDwtSFg7R1kTQBXVJA4Nn84ABMKyuimXQKmHTgAMQhJNiYYC11kECYCmAhjEAALGDZCAHUR4nhSbLA4NTdZ4gA3WYFesFI-EG0UyRgxIJ0OxOTigCtSejVcwjgARQKIeAvCFJNs6SlenSCQhGAEdCMMhFpYJNiBUYCNkAAOAAM3wcKUmgN6ELghzW8RMwNe8BOZncfVIeCgcEqlWGkjghOI9JOjXOSCSn2KKTYxHykiKJlID2eRPi7MuJhgTH80Nh8KQACZJUFiCRdABhCAcJi9HEAVmGBRSABVpe4OV8QGsigBJKDVWBqMBacwAQXtGgEjw+KX0+iAA)

```sh
--parser graphql
```

**Input:**

```graphql
"""
  - list
  - list
extea comment
"""
type Person {
  name: String
}
```

**Output:**

```graphql
"""
- list
  - list
extea comment
"""
type Person {
  name: String
}

```

File: src/language-graphql/printer-graphql.js (hunk 1 above)
Comment (chimurai) [reply src/language-graphql/printer-graphql.js:131]: There is spec:
https://spec.graphql.org/October2021/#sec-Descriptions

> GraphQL descriptions are defined using the Markdown syntax (as specified by [CommonMark](https://commonmark.org/)).

File: src/language-graphql/printer-graphql.js [hunk 2]
```diff
@@ -128,7 +128,12 @@ function genericPrint(path, options, print) {
...
+          join(hardline,
+            node.value
+            .replace(/"""/g, "\\$&")
+            .split("\n")
+            .map((val, idx, arr) => arr.length === 1 ? val.trim() : val)
```
Comment (chimurai) [inline src/language-graphql/printer-graphql.js:135]: Only trim value when it's a single line description, so it doesn't affect multi-line descriptions

File: src/language-graphql/printer-graphql.js (hunk 1 above)
Comment (chimurai) [reply src/language-graphql/printer-graphql.js:131]: Added `Markdown` support in GraphQL `Descriptions`. This should resolve the weird use cases you encountered with jumping leading spaces.

**Comment by chimurai** [comment]
Getting this error on [deploy-preview-12608](https://deploy-preview-12608--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAdNGAEBNT79qwCeADnJgBqbCpSaZQCGAtnEpgMowBOAllAHNaAXxAAaEBBIxe0AM7JQjbtwgB3AArKEClIwA2axkQUSARt0ZgA1nBgcWcADL84yAGYG5cc5Zt2OEit+AWQeAFcfEDhmMzgAE3iEp0ZBcMYBOAAxCG5mRhgZQWQQRnCYCHEQAAsYZn0AdWreeDkgsDgOHRbeADcWohKwOVMQfm9uGA1LAXyPLyiAKzkADw4Q-TgARXCIeHn9bwkg7gmSgUsSaoBHfSqSPlgG3niYauQADgAGY9VvBsuJQecAmvTcEmuu3g0ykulKcgAtFA4AkElVuHBIbwMdMMnMkJ5DlFvMxeGFuJEJHINtsoW4CQsJDBGGZnq93kgAExMyy8fQhADCEGY+OicgArFVwt4ACos3SEo4gXqRACSUCSsA4YD40gAghquERNgdvMJhEA):
> Error: Couldn't resolve parser "markdown". Parsers must be explicitly added to the standalone bundle.

Not sure how to solve this.
From unit test point of view, the Markdown support seems to look ok.

File: src/language-graphql/printer-graphql.js [hunk 3]
```diff
@@ -125,10 +126,18 @@ function genericPrint(path, options, print) {
     }
     case "StringValue": {
       if (node.block) {
+        const plugins = [
+          require("../language-markdown/index.js"),
+          require("../language-graphql/index.js"),
+          require("../language-js/index.js"),
+        ];
+        const formatMarkdown = (v) => formatWithCursor(v, { plugins, parser: "markdown" });
+        const { formatted } = formatMarkdown(node.value.replace(/"""/g, "\\$&"));
```
Comment (fisker) [inline src/language-graphql/printer-graphql.js:135]: This is not the right way to fo call anther parser. I suggest you focus on fixing #12608.

File: changelog_unreleased/graphql/12608.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+# Input
+""" Customer """
+type Person {
+  name: String
+}
+
+# Prettier 2.6.2
+"""
+ Customer
+"""
```
Comment (fisker) [inline changelog_unreleased/graphql/12608.md:14]: Why don't we try print the comment inline, if:

- can fit
- was inline originally

**Review by fisker — APPROVED** [review]

**Comment by chimurai** [comment]
Sounds like a good approach to deal with the single line description first.

Seeing the same failure in `Dev / Node.js 16 on ubuntu-latest (pull_request)` (Previous run https://github.com/prettier/prettier/pull/12608#issuecomment-1091348080)

**Comment by fisker** [comment]
I didn't notice that, we can fix this here https://github.com/prettier/prettier/blob/c50f2bbe3075757840bbd7ed941647a9bde84266/src/language-graphql/printer-graphql.js#L575, we need apply the same logic there, you can check how `clean` works in other language.

**Comment by fisker** [comment]
Use `FULL_TEST=1 yarn test` to test on your machine.

**Comment by chimurai** [comment]
Added logic to the `clean()` function.
Let me know if it needs adjustment.

File: src/language-graphql/printer-graphql.js [hunk 5]
```diff
@@ -572,7 +574,11 @@ function printInterfaces(path, options, print) {
   return parts;
 }
 
-function clean(/*node, newNode , parent*/) {}
+function clean(node, newNode /* , parent */) {
+  if (node.kind === "StringValue" ) {
```
Comment (fisker) [inline src/language-graphql/printer-graphql.js:578]:
```suggestion
    if (node.kind === "StringValue" && node.block && !node.value.includes("\n")) {
```

File: src/language-graphql/printer-graphql.js [hunk 6]
```diff
@@ -572,7 +574,11 @@ function printInterfaces(path, options, print) {
...
+    newNode.value = newNode.value.trim();
```
Comment (fisker) [inline src/language-graphql/printer-graphql.js:579]:
```suggestion
    newNode.value = newNode.value.trimStart();
```

File: src/language-graphql/printer-graphql.js [hunk 7]
```diff
@@ -125,13 +125,15 @@ function genericPrint(path, options, print) {
     }
     case "StringValue": {
       if (node.block) {
-        return [
-          '"""',
-          hardline,
-          join(hardline, node.value.replace(/"""/g, "\\$&").split("\n")),
+        const lines = node.value.replace(/"""/g, "\\$&").split("\n");
+        if (lines.length === 1) {
+          lines[0] = lines[0].trim();
```
Comment (fisker) [inline src/language-graphql/printer-graphql.js:130]:
```suggestion
          lines[0] = lines[0].trimStart();
```

**Comment by Harry-Hopkinson** [comment]
Now you just need to lint the file

File: src/language-graphql/printer-graphql.js (hunk 5 above)
Comment (Harry-Hopkinson) [inline src/language-graphql/printer-graphql.js:578]:
```suggestion
  if (node.kind === "StringValue") {
```

Comment (Harry-Hopkinson) [reply src/language-graphql/printer-graphql.js:578]: This will remove the failing linting task.

**Review by Harry-Hopkinson — CHANGES_REQUESTED** [review]

File: src/language-graphql/printer-graphql.js (hunk 5 above)
Comment (Harry-Hopkinson) [reply src/language-graphql/printer-graphql.js:578]: I do not have permission to push to the repo so I can fix the task at this period of time

**Review by Harry-Hopkinson — APPROVED** [review]

**Comment by fisker** [comment]
https://github.com/prettier/prettier/pull/12608#discussion_r850012263

File: src/language-graphql/printer-graphql.js (hunk 5 above)
Comment (chimurai) [reply src/language-graphql/printer-graphql.js]: Included the suggestion with comments. (Makes more sense indeed)

**Comment by chimurai** [comment]
https://github.com/prettier/prettier/pull/12608#discussion_r850012263 has been fixed.

**Review by Harry-Hopkinson — APPROVED** [review]
Looks Good

**Review by sosukesuzuki — APPROVED** [review]

**Pull request merged**